src/java/org/apache/cassandra/cql3/restrictions/SingleColumnRestriction.java

```diff
@@ -440,6 +442,139 @@ private SliceRestriction(ColumnMetadata columnDef, TermSlice slice)
         }
     }
 
+    /**
+     * One or more slice restrictions on a column's map entries.
+     * For a map column of type map&lt;text,int&gt; with name m, here are some examples of valid restrictions:
+     * One restriction:                m['a'] > 1
+     * Restrictions on different keys: m['a'] > 1 AND m['b'] < 2
+     * Restrictions on same key:       m['a'] > 0 AND m['a'] < 2
+     */
+    public static class MapSliceRestriction extends SingleColumnRestriction
+    {
+        // Left is the map's key and right is the slice on the map's value.
+        private final List<Pair<Term, TermSlice>> slices;
+
+        public MapSliceRestriction(ColumnMetadata columnDef, Bound bound, boolean inclusive, Term key, Term value)
+        {
+            super(columnDef);
+            slices = new ArrayList<>();
+            slices.add(Pair.create(key, TermSlice.newInstance(bound, inclusive, value)));
+        }
+
+        private MapSliceRestriction(ColumnMetadata columnDef, List<Pair<Term, TermSlice>> slices)
+        {
+            super(columnDef);
+            this.slices = slices;
+        }
+
+        @Override
+        public void addFunctionsTo(List<Function> functions)
+        {
+            slices.forEach(slice -> {
+                slice.left.addFunctionsTo(functions);
+                slice.right.addFunctionsTo(functions);
+            });
+        }
+
+        @Override
+        MultiColumnRestriction toMultiColumnRestriction()
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public MultiCBuilder appendTo(MultiCBuilder builder, QueryOptions options)
+        {
+            // MapSliceRestrictions are not supported on clustering columns.
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public boolean hasBound(Bound b)
+        {
+            // Because a MapSliceRestriction can have multiple slices, we cannot implement this method.
+            throw new UnsupportedOperationException("Bounds not well defined for map slice restrictions");
+        }
+
+        @Override
+        public boolean isInclusive(Bound b)
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public SingleRestriction doMergeWith(SingleRestriction otherRestriction)
+        {
+            checkTrue(otherRestriction instanceof SingleColumnRestriction.MapSliceRestriction,
+                      "Column \"%s\" cannot be restricted by both an inequality relation and \"%s\"",
```

Comment:
```suggestion
                      "Column \"%s\" cannot be restricted by both map entry comparison relation and \"%s\"",
```

Not sure what a user friendly working is here, but "inequality" seems like it is a copy paste error :)

Reply:
I was thinking of the mathematical definition of inequality, which likely fits, but perhaps isn't the right term in the db context. Your error message works too. I'll switch.

Reply:
FWIW, we have used `inequality` in the `SliceRestriction` exception since 2014. Unless you feel strongly that it's wrong, I think we can stick with it here too.

Reply:
Ah. Got it. No that makes sense.